main/network.c

```diff
@@ -825,88 +822,80 @@ php_socket_t php_network_connect_socket_to_host(const char *host, unsigned short
 	for (sal = psal; !fatal && *sal != NULL; sal++) {
 		sa = *sal;
 
-		/* create a socket for this address */
-		sock = socket(sa->sa_family, socktype, 0);
-
-		if (sock == SOCK_ERR) {
-			continue;
-		}
-
 		switch (sa->sa_family) {
 #if HAVE_GETADDRINFO && HAVE_IPV6
 			case AF_INET6:
 				if (!bindto || strchr(bindto, ':')) {
-					((struct sockaddr_in6 *)sa)->sin6_family = sa->sa_family;
 					((struct sockaddr_in6 *)sa)->sin6_port = htons(port);
 					socklen = sizeof(struct sockaddr_in6);
 				} else {
-					socklen = 0;
-					sa = NULL;
+					/* Expect IPV4 address, skip to the next */
+					continue;
 				}
 				break;
 #endif
 			case AF_INET:
-				((struct sockaddr_in *)sa)->sin_family = sa->sa_family;
 				((struct sockaddr_in *)sa)->sin_port = htons(port);
 				socklen = sizeof(struct sockaddr_in);
+				if (bindto && strchr(bindto, ':')) {
+					/* IPV4 sock can not bind to IPV6 address */
+					bindto = NULL;
+				}
 				break;
 			default:
-				/* Unknown family */
-				socklen = 0;
-				sa = NULL;
+				/* Unsupported family, skip to the next */
+				continue;
+		}
+
+		/* create a socket for this address */
+		sock = socket(sa->sa_family, socktype, 0);
+
+		if (sock == SOCK_ERR) {
+			continue;
 		}
 
 		if (sa) {
 			/* make a connection attempt */
 
 			if (bindto) {
-				struct sockaddr *local_address = NULL;
-				int local_address_len = 0;
+				struct {
+					int len;
+					union {
+						struct sockaddr common;
+						struct sockaddr_in in4;
+#if HAVE_IPV6 && HAVE_INET_PTON
+						struct sockaddr_in6 in6;
+#endif
+					};
+				} local_address;
 
+				local_address.len = 0;
 				if (sa->sa_family == AF_INET) {
-					if (strchr(bindto,':')) {
-						goto skip_bind;
-					}
-					struct sockaddr_in *in4 = emalloc(sizeof(struct sockaddr_in));
-
-					local_address = (struct sockaddr*)in4;
-					local_address_len = sizeof(struct sockaddr_in);
-
-					in4->sin_family = sa->sa_family;
-					in4->sin_port = htons(bindport);
 #ifdef HAVE_INET_PTON
-					if (!inet_pton(AF_INET, bindto, &in4->sin_addr)) {
```

Comment:
The check ways of the two `inet_pton` calls are not consistent, I think it might be better to judge whether it returns 1